ui/versionmanager/version_manager.go

```diff
@@ -0,0 +1,289 @@
+func NewVersionManager(
+	uiServer NodeUIServer,
+	http *requests.HTTPClient,
+	versionConfig *VersionConfig,
+) *VersionManager {
+	return &VersionManager{
+		uiServer:      uiServer,
+		httpClient:    http,
+		versionConfig: versionConfig,
+		github:        newGithub(http),
+		downloader:    NewDownloader(),
+	}
+}
+
+// TODO check integrity of downloaded release so not to serve a broken nodeUI
+
+// ListLocalVersions list downloaded Node UI versions
+func (vm *VersionManager) ListLocalVersions() ([]LocalVersion, error) {
+	files, err := ioutil.ReadDir(vm.versionConfig.nodeUIDir)
+	if err != nil {
+		return nil, fmt.Errorf("could not read "+nodeUIPath+": %w", err)
+	}
+
+	var versions []LocalVersion
+	for _, f := range files {
+		if f.IsDir() {
+			versions = append(versions, LocalVersion{
+				Name: f.Name(),
+			})
+		}
+	}
+
+	return versions, nil
+}
+
+// UsedVersion current version
+func (vm *VersionManager) UsedVersion() (LocalVersion, error) {
+	version, err := vm.versionConfig.Version()
+	if err != nil {
+		return LocalVersion{}, err
+	}
+	return LocalVersion{
+		Name: version,
+	}, nil
+}
+
+// BundledVersion bundled version
+func (vm *VersionManager) BundledVersion() (LocalVersion, error) {
+	name, err := godvpnweb.Version()
+	if err != nil {
+		return LocalVersion{}, err
+	}
+	return LocalVersion{
+		Name: name,
+	}, nil
+}
+
+// RemoteVersionRequest for paged requests
+type RemoteVersionRequest struct {
+	PerPage    int
+	Page       int
+	FlushCache bool
+}
+
+// ListRemoteVersions list versions from github releases of NodeUI
+func (vm *VersionManager) ListRemoteVersions(r RemoteVersionRequest) ([]RemoteVersion, error) {
+	if time.Now().Before(vm.remoteCacheExpiresAt) && vm.releasesCache != nil && !r.FlushCache {
+		return remoteVersions(vm.releasesCache), nil
+	}
+
+	releases, err := vm.github.nodeUIReleases(r.PerPage, r.Page)
+	if err != nil {
+		return nil, err
+	}
+
+	vm.releasesCache = releases
+	vm.remoteCacheExpiresAt = time.Now().Add(time.Hour)
+
+	return remoteVersions(vm.releasesCache), nil
+}
+
+func remoteVersions(releases []GitHubRelease) []RemoteVersion {
+	var versions []RemoteVersion
+	for _, release := range releases {
+		versions = append(versions, RemoteVersion{
+			Name:             release.Name,
+			PublishedAt:      release.PublishedAt,
+			CompatibilityURL: compatibilityAssetURL(release),
+			IsPreRelease:     release.Prerelease,
+		})
+	}
+	return versions
+}
+
+func compatibilityAssetURL(r GitHubRelease) string {
+	for _, a := range r.Assets {
+		if a.Name == compatibilityAssetName {
+			return a.BrowserDownloadUrl
+		}
+	}
+	return ""
+}
+
+// TODO think about sending SSE to inform to nodeUI that a version has been downloaded
+
+// Download download and untar node UI dist
+func (vm *VersionManager) Download(versionName string) error {
+	assetURL, err := vm.github.nodeUIDownloadURL(versionName)
+	if err != nil {
+		return err
+	}
+
+	err = os.MkdirAll(vm.versionConfig.uiDistPath(versionName), 0700)
```

Comment:
`DownloadNodeUI` Might be already downloading this version or it might exist, but you will still attempt to create a dir for it now.

Reply:
Does not cause error if path already exists